src/WebJobs.Script/IO/AutoRecoveringFileSystemWatcher.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host;
+
+namespace Microsoft.Azure.WebJobs.Script.IO
+{
+    public class AutoRecoveringFileSystemWatcher : IDisposable
+    {
+        private readonly string _path;
+        private readonly string _filter;
+        private readonly bool _includeSubdirectories;
+        private readonly WatcherChangeTypes _changeTypes;
+        private readonly TraceWriter _traceWriter;
+        private readonly Action<ErrorEventArgs> _handleFileError;
+        private readonly CancellationToken _cancellationToken;
+        private readonly object _syncRoot = new object();
+        private CancellationTokenSource _cancellationTokenSource;
+        private FileSystemWatcher _fileWatcher;
+        private bool _disposed = false;
+        
+        public event EventHandler<FileSystemEventArgs> Changed;
+
+        public AutoRecoveringFileSystemWatcher(string path, string filter = "*.*", 
+            bool includeSubdirectories = true, WatcherChangeTypes changeTypes = WatcherChangeTypes.All, TraceWriter traceWriter = null)
+        {
+            _path = path;
+            _filter = filter;
+            _changeTypes = changeTypes;
+            _includeSubdirectories = includeSubdirectories;
+            _traceWriter = traceWriter;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _cancellationToken = _cancellationTokenSource.Token;
+            _handleFileError = new Action<ErrorEventArgs>(OnFileWatcherError).Debounce();
+            
+            InitializeWatcher();
+        }
+      
+        ~AutoRecoveringFileSystemWatcher()
+        {
+            Dispose(false);
+        }
+
+        private void InitializeWatcher()
+        {
+            if (!Directory.Exists(_path))
+            {
+                throw new DirectoryNotFoundException($"The path '{_path}' cannot be found.");
+            }
+
+            lock (_syncRoot)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                _fileWatcher = new FileSystemWatcher(_path, _filter)
+                {
+                    IncludeSubdirectories = _includeSubdirectories,
+                    EnableRaisingEvents = true
+                };
+
+                _fileWatcher.Error += (s, a) => _handleFileError(a);
+                AddEventSubscriptions(_fileWatcher, _changeTypes);
+            }
+        }
```

Comment:
It will (unfortunately) only get the events from the recovery point on. 

In order to detect changes that happened while in a bad state, we'd have to implement that logic.